### DISCUSSION TITLE
TextGrad integration

### DISCUSSION THREAD

**Discussion by sakthi-geek** [langfuse/langfuse#3478] [Ideas]
### Describe the feature or potential improvement

TextGrad is a Python package that provides a simple interface to implement LLM-“gradients” pipelines for text optimization!

Check out the [Usage](https://textgrad.readthedocs.io/en/latest/usage.html) section for further information, including how to install the project. Want to directly jump to the optimization process? Check out the [QuickStart](https://textgrad.readthedocs.io/en/latest/quickstart.html) guide!

An autograd engine – for textual gradients!

TextGrad is a powerful framework building automatic [``](https://textgrad.readthedocs.io/en/latest/#id1)differentiation’’ via text. TextGrad implements backpropagation through text feedback provided by LLMs, strongly building on the gradient metaphor.

We provide a simple and intuitive API that allows you to define your own loss functions and optimize them using text feedback. This API is similar to the Pytorch API, making it simple to adapt to your use cases.

Paper - [https://arxiv.org/abs/2406.07496](https://arxiv.org/abs/2406.07496)

Github - [https://github.com/zou-group/textgrad](https://github.com/zou-group/textgrad)

**Comment by marcklingen** [comment]
Looks cool, thanks for sharing!

It seems like [`textgrad.engine.base`](https://textgrad.readthedocs.io/en/latest/modules/textgrad.engine.base.html) wraps different models. If you want to add logging to Langfuse to TextGrad you could implement it in a very similar way as mirascope did it ([docs](https://docs.mirascope.io/latest/integrations/langfuse/?h=langfuse)) as they just reused the langfuse python decorator.
